### PR TITLE
 Automatic tag table generation

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
Without this change we don't even really know how many tags are created and what are their values.
With this change we want this information to be automatically rendered.

### CODE REVIEW

File: spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/TagKey.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	 * @return tag key
+	 */
+	String getKey();
+
+	/**
+	 * Merges arrays of tags.
+	 * @param tags array of tags
+	 * @return a merged array of tags
+	 */
+	static TagKey[] merge(TagKey[]... tags) {
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/TagKey.java:41]: This is basically  `concat`, right? If so I think I would rename it, also, if I understand this correctly, this should do the same in a simpler way:
```java
return Arrays.stream(tags)
    .flatMap(Arrays::stream)
    .toArray(TagKey[]::new);
```

**Review by jonatan-ivanov — APPROVED** [review]

File: spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/DocumentedSpanAssertions.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	static boolean SLEUTH_SPAN_ASSERTIONS_ON = Boolean.parseBoolean(System.getProperty(
+			"spring.cloud.sleuth.assertions.enabled", System.getenv("SPRING_CLOUD_SLEUTH_ASSERTIONS_ENABLED") != null
+					? System.getenv("SPRING_CLOUD_SLEUTH_ASSERTIONS_ENABLED") : "false"));
+
+	private static final Map<String, Pattern> PATTERN_CACHE = new ConcurrentHashMap<>();
+
+	private static final Pattern SPECIAL_REGEX_CHARS = Pattern.compile("[{}()\\[\\].+*?^$\\\\|]");
+
+	private DocumentedSpanAssertions() {
+		throw new IllegalStateException("Can't instantiate utility class");
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/DocumentedSpanAssertions.java:43]: I have those `ISE` everywhere else in such cases so I'll leave this one here too.

File: spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/DocumentedSpan.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	/**
+	 * @return span name
+	 */
+	String getName();
+
+	/**
+	 * @return allowed tag keys
+	 */
+	default TagKey[] getTagKeys() {
```
[the remark this answers is not in the document]
Comment (marcingrzejszczak) [reply spring-cloud-sleuth-api/src/main/java/org/springframework/cloud/sleuth/docs/DocumentedSpan.java:50]: Not really cause `SpanEntry` is just a pojo that I'm using in the table. It's much easier to work with `TagKey[]` when enums in their `.values()` method return an array.

**Comment by marcingrzejszczak** [comment]
Done via https://github.com/spring-cloud/spring-cloud-sleuth/commit/d2041cbd47ca7cd232b04530a9cf9e468bd3fb40

**Pull request closed without merging**